mmdet3d/models/detectors/imvotenet.py

```diff
@@ -0,0 +1,764 @@
+            calibs (dict[str, torch.Tensor]): camera calibration matrices,
+                Rt and K.
+            bboxes_2d (list[torch.Tensor]): provided 2d bboxes,
+                not supported yet.
+            gt_bboxes_3d (:obj:`BaseInstance3DBoxes`): 3d gt bboxes.
+            gt_labels_3d (list[torch.Tensor]): gt class labels for 3d bboxes.
+            pts_semantic_mask (None | list[torch.Tensor]): point-wise semantic
+                label of each batch.
+            pts_instance_mask (None | list[torch.Tensor]): point-wise instance
+                label of each batch.
+
+        Returns:
+            dict[str, torch.Tensor]: a dictionary of loss components.
+        """
+
+        if points is None:
+            x = self.extract_img_feat(img)
+            losses = dict()
+
+            # RPN forward and loss
+            if self.with_img_rpn:
+                proposal_cfg = self.train_cfg.get('img_rpn_proposal',
+                                                  self.test_cfg.img_rpn)
+                rpn_losses, proposal_list = self.img_rpn_head.forward_train(
+                    x,
+                    img_metas,
+                    gt_bboxes,
+                    gt_labels=None,
+                    gt_bboxes_ignore=gt_bboxes_ignore,
+                    proposal_cfg=proposal_cfg)
+                losses.update(rpn_losses)
+            else:
+                proposal_list = proposals
+
+            roi_losses = self.img_roi_head.forward_train(
+                x, img_metas, proposal_list, gt_bboxes, gt_labels,
+                gt_bboxes_ignore, gt_masks, **kwargs)
+            losses.update(roi_losses)
+            return losses
+        else:
+            with torch.no_grad():
+                bboxes_2d = self.extract_bboxes_2d(
+                    img, img_metas, bboxes_2d=bboxes_2d, **kwargs)
+
+            points = torch.stack(points)
+            seeds_3d, seed_3d_features, seed_indices = \
+                self.extract_pts_feat(points)
+
+            img_features, masks = self.fusion_layer(img, bboxes_2d, seeds_3d,
+                                                    img_metas, calibs)
+
+            inds = sample_valid_seeds(masks, self.num_sampled_seed)
+            batch_size, img_feat_size = img_features.shape[:2]
+            pts_feat_size = seed_3d_features.shape[1]
+            inds_img = inds.reshape(batch_size, 1,
+                                    -1).repeat(1, img_feat_size, 1)
+            img_features = img_features.gather(-1, inds_img)
+            inds = inds % inds.shape[1]
+            inds_seed_xyz = inds.reshape(batch_size, -1, 1).repeat(1, 1, 3)
+            seeds_3d = seeds_3d.gather(1, inds_seed_xyz)
+            inds_seed_feats = inds.reshape(batch_size, 1,
+                                           -1).repeat(1, pts_feat_size, 1)
+            seed_3d_features = seed_3d_features.gather(-1, inds_seed_feats)
+            seed_indices = seed_indices.gather(1, inds)
```

Comment:
Can we use view/expand to replace reshape/repeat to reduce memory costs? Or do we need? (Do these intermediate outputs have large shape?)

Reply:
According to PyTorch doc, reshape is equivalent with view when input is contiguous, so using reshape should be OK.

Reply:
As for repeat v.s. expand, yes, we should use expand.

Reply:
Nevertheless, I'll change all reshapes to views if possible